tools/jvm/src/main/scala/org/scalajs/core/tools/io/AtomicFileOutputStream.scala

```diff
@@ -0,0 +1,37 @@
+package org.scalajs.core.tools.io
+
+import java.io._
+import java.net.URI
+
+/** Provides best-effort atomic writing to a file
+ *
+ *  This FileOutputStream writes to a temporary file and moves it onto
+ *  the original file, once this FileOutputStream is closed. However,
+ *  it is only able to do so, if the underlying filesystem supports
+ *  it. If it doesn't it falls back to non-atomic moving or copying.
+ */
+private[io] class AtomicFileOutputStream private (
+    private val baseFile: File,
+    private val tmpFile: File
+) extends FileOutputStream(tmpFile) {
+
+  def this(baseFile: File) = {
+    this(baseFile, {
+      // Create a temporary file we actually write to
+      val tmpFile = File.createTempFile(
+          ".tmp-", baseFile.getName, baseFile.getParentFile)
```

Comment:
I wanted ".tmp-" to be the prefix in an attempt to hide the file. I put it in the same directory as the target file, since otherwise it might lay across an FS boundary which will jeopardize atomically moving the file.

This is only a best-effort heuristics, there is no platform-independent way to do that in Java AFAIK (after research).